### PR TITLE
Bug 1494666 - Choose to open new tab in foreground

### PR DESCRIPTION
**Pull request by blewjy**
This PR fixes [Bug 1494666](https://bugzilla.mozilla.org/show_bug.cgi?id=1494666).

I have added a new `UISwitch` to the "New Tab Settings" that allows the user to select if they wish to open all new tabs in background or foreground. If tabs are opened in the foreground, the browser will immediately switch to that new tab. If tabs are opened in the background, browser will show a toast message signifying that the new tab has been opened (basically similar behaviour to what was before this PR).

It might be worth to see if we can add some animations when a new tab is opened in the foreground, similar to Safari where all tabs are momentarily displayed before the new tab is focused. For now, the new tab is simply immediately focused, similar to switching tabs from the toast message.

Note that this change **does not** affect the browsers with the top tabs visible, namely the larger devices (iPad).

Screenshots available below.

## Pull Request Checklist

- [x] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Bug 12345678 - This fixes something something`

- [x] I have updated the *Unit Tests* to cover new or changed functionality
- [x] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [x] I have made sure that localizable strings use `NSLocalizableString()`

### Before:
bug-1494666-before-settings
bug-1494666-before

### After:
bug-1494666-after-settings

### After (foreground):
bug-1494666-after-foreground

### After (background):
bug-1494666-after-background

### CODE REVIEW

**Comment by bryanbell** [comment]
This is a good and useful feature, but the UI is in the wrong place. Can you place this feature in a new subsection of General called Tabs? 

image

**Comment by farhanpatel** [comment]
Thanks for the PR. You hit the nail on the head with the NSLocalizedStrings. Could you please move them into Strings.swift.

**Comment by blewjy** [comment]
Thank you for your guidance, @farhanpatel! I have updated the PR accordingly :)

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -2227,6 +2216,26 @@ extension BrowserViewController: ContextMenuHelperDelegate {
     }
 }
 
+extension BrowserViewController {
+    func checkIfSwitchTabImmediately(tab: Tab) {
+        // Depending on user preference, new tab to be opened in background or be switched to immediately
+        // Default is to open in foreground
+        let switchToNewTabImmediately = self.profile.prefs.boolForKey(PrefsKeys.KeySwitchToNewTabImmediately) ?? true
```
Comment (farhanpatel) [inline Client/Frontend/Browser/BrowserViewController.swift]: The current default behaviour is to open the tab in the background and show the ButtonToast. Can this be `?? false` instead.

File: Client/Frontend/Settings/TabSettingsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        self.title = Strings.SettingsTabsTitle
+        hasSectionSeparatorLine = false
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func generateSettings() -> [SettingSection] {
+        let switchToNewTabSetting = BoolSetting(prefs: profile.prefs, prefKey: PrefsKeys.KeySwitchToNewTabImmediately, defaultValue: true, attributedTitleText: NSAttributedString(string: Strings.SettingsTabsImmediateSwitch))
```
Comment (farhanpatel) [inline Client/Frontend/Settings/TabSettingsViewController.swift]: The default value here should be changed here as well.

File: package-lock.json [hunk 3]
```diff
@@ -3741,8 +3741,9 @@
       "dev": true
     },
     "page-metadata-parser": {
-      "version": "github:mozilla/page-metadata-parser#66397fbafb5aab3a7840854d9d3a79c72a9620e6",
-      "from": "github:mozilla/page-metadata-parser#66397fbafb5aab3a7840854d9d3a79c72a9620e6"
+      "version": "1.1.3",
+      "resolved": "https://registry.npmjs.org/page-metadata-parser/-/page-metadata-parser-1.1.3.tgz",
+      "integrity": "sha512-/ILdoE7T62QAqUllDLuZ7mA0x7yYc4S94jjgvz5vxqhYLU2BCUp20SxVir0gq3kezXXAsfLg768UkbvvqYeJ5g=="
```
Comment (farhanpatel) [inline package-lock.json]: I'm pretty sure this is a valid change but for the sake of this PR can you please remove the changes to this file.

**Comment by blewjy** [comment]
Looking for anyone to help take a look at this PR!

It seems that the build is failing due to the change in `package-lock.json`...

**Comment by blewjy** [comment]
It has been a whole year lol. Closing this

**Pull request closed without merging**